Bdms 227 transfer updates

jacob-a-brown opened the pull request:
all transfers for additional well information except for aquifer/geology data. this depends on PR #256 being merged since it incorporates api/model/schema changes made for the tests to pass

jacob-a-brown committed:
feat: transfer water level sample permissions

jacob-a-brown committed:
feat: transfer permissions from legacy db

jacob-a-brown committed:
feat: transfer well construction information

jacob-a-brown committed:
WIP: transfers for well additional info

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
feat: transfer well source information

jacob-a-brown committed:
feat: add Quemado Mutual Water and Sewage Works Association to organizations for contacts

jacob-a-brown committed:
fix: remove erroneous or check

jacob-a-brown committed:
feat: transfer well pump type

parse the construction notes field to get pump types. this can be more
sophisticated, but that's a future problem. for now just check for some
basic pump type string patterns

jacob-a-brown committed:
fix: change Air-rotary to Air-Rotary to correspond with lu table

This can be updated later on but should be "Air-Rotary" for the transfer

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
fix: remove duplicate well pump types in lexicon

this was an artifact from merge conflicts

jacob-a-brown committed:
fix: use OpenWellLoggerOK for is_suitable_for_datalogger field

Laila indicated that this just describes if a datalogger can be
physically installed at the well. It does not pertain to
permissions

jacob-a-brown committed:
feat: update CreateWell for transfers

ksmuczynski commented on db/thing.py:
```diff
@@ -484,6 +545,10 @@ class WellScreen(Base, AutoBaseMixin, ReleaseMixin):
         "AquiferSystem", back_populates="well_screens", passive_deletes=True
     )
 
+    geologic_formation: Mapped["GeologicFormation"] = relationship(
+        "GeologicFormation", back_populates="well_screens", passive_deletes=True
+    )
+
```
Do these relationships belong in the `ThingIdLink` model?

jacob-a-brown replied:
I don't think that they would. These two are in the `WellScreen` model

ksmuczynski replied:
Oh, I thought I saw them in the ThingIdLink model, which is why I was confused. All good!

ksmuczynski commented on schemas/thing.py:
```diff
@@ -24,18 +24,20 @@
+
+# from schemas.geologic_formation import (
+#     GeologicFormationResponse,
+# )
```
Don't we still need a GeologicFormationResponse?

jacob-a-brown replied:
I don't think we do. The feature file states that it should just return the formation codes, so I parse all of the geologic formation and just return the codes as plain strings

ksmuczynski commented on schemas/thing.py:
```diff
@@ -24,18 +24,20 @@
-    GeologicFormationResponse,
-    ThingGeologicFormationAssociationResponse,
-)
+from schemas.aquifer_system import AquiferSystemGeoJSONResponse
```
I thought we decided to forego the GeoJSON response?

jacob-a-brown replied:
And I'll just include the `name` of hte aquifer system in the `WellScreenResponse`. It's not currently used, but that shouldn't be an issue for the time being

jacob-a-brown replied:
This was used in the `WellResponse` and I forgot it was there. I've now changed it so that a string of the aquifer system name is returned

ksmuczynski commented on schemas/thing.py:
```diff
@@ -229,18 +238,23 @@ class WellResponse(BaseThingResponse):
+    allow_water_level_samples: bool | None
+    allow_water_chemistry_samples: bool | None
+    allow_datalogger_installation: bool | None
+    is_suitable_for_datalogger: bool | None
```
These look like fields from the previous PermissionHistory table design. Should these fields be removed and replaced with `permission_type` and/or `permission_allowed`?

jacob-a-brown replied:
Sorry if you are getting this twice. I somehow started a review of my own PR, which I just deleted so now I need to comment again:

The steps in the feature file are

```
And the response should include whether repeat measurement permission is granted for the well
And the response should include whether sampling permission is granted for the well
And the response should include whether datalogger installation permission is granted for the well
```

so I'm just returning boolean values based off of the latest response where `end_date` is `None`. This data is, however, stored in the `PermisionHistory` table.

@chasetmartin do you want the response to be single fields like this, or would you prefer that it more closely reflect the database design? If so I can change it to

```python
{
...
"permissions": [
  {"permission_type": "Water Level Sample", "permission_allowed": True/False},
  {"permission_type": "Water Chemistry Sample", "permission_allowed": True/False},
  {"permission_type": "Datalogger Installation", "permission_allowed": True/False}
]
}

chasetmartin replied:
@jacob-a-brown I like the later, the list of permission objects, I think it's almost more intuitive to read/parse. What do you think @TylerAdamMartinez ?

TylerAdamMartinez replied:
Same here, let's make it more closely resemble the database design.

jacob-a-brown replied:
Do you want it to return _all_ permissions, even if they have ended? Or just the latest permissions (that is, latest `start_date` where `end_date` is `None`)?

jacob-a-brown replied:
the permissions are now returned in a list

```python
{
...
"permissions": [
  {"permission_type": "Water Level Sample", "permission_allowed": True/False, "start_date": date, "end_date": None},
  {"permission_type": "Water Chemistry Sample", "permission_allowed": True/False, "start_date": date, "end_date": None},
  {"permission_type": "Datalogger Installation", "permission_allowed": True/False, "start_date": date, "end_date": None}
]
}
```

- Only the latest record (based off of `start_date`) where `end_date` is `None` is returned for each type
- If the permission record does not exist for a particular `permission_type`then `permission_allowed`, `start_date`, and `end_date` are `None`.
- while `start_date` is non-nullable in the database, it must be nullable in the response incase there are no permissions of  particular type
- `end_date` will always be `None` for active permissions. I'm including it in the `PermissionHistory` response, though, so that if we want to list out the full permission history for an object it'll be there

ksmuczynski reviewed (commented):
Overall looks good! Left a few comments.

jacob-a-brown committed:
fix: return aquifer/geology str instead of full response

the well screen response includes the aquifer and geology information,
but just as strings of names and codes instead of the full nested response objects.

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-transfer-updates

ksmuczynski reviewed (approved):
Nice work!
